barter/examples/custom_strategy.rs

```diff
@@ -0,0 +1,414 @@
+        }
+
+        // There are already some opened orders. Skip the current evaluation.
+        // This is here so that we don't push many buy orders for the same
+        // desired quantity, before even receiving an event of execution.
+        // TODO: How should this atomicity actually be handled? There are cases
+        // when we can have orders as marked completed and no trade events yet
+        // received from the exchange.
+        let mut current_orders = state
+            .instruments
+            .instrument_index(&self.instrument)
+            .orders
+            .orders();
+        if current_orders.any(|o| o.state.is_open_or_in_flight()) {
+            return (vec![], vec![]);
+        }
+
+        // Instrument data we are trading
+        let instrument = state.instruments.instrument_index(&self.instrument);
+
+        // Current market price. This is needed because of how the MockExchange
+        // currently mocks order executions. Ideally we wouldn't need to pass
+        // this in the future.
+        // TODO: Ugly type annotation needed.
+        let Some(price) =
+            <DefaultMarketData as MarketDataState<InstrumentIndex>>::price(&instrument.market)
+        else {
+            warn!("Market price is not yet set");
+            return (vec![], vec![]);
+        };
+
+        // Available quote balance for the instrument
+        let available_quote = state
+            .assets
+            .asset_index(&instrument.instrument.underlying.quote)
+            .balance
+            .expect("balance should exist for the quote asset")
+            .value
+            .free;
+
+        info!(%available_quote, "Available quote amount");
+
+        // Buy order we are proposing
+        let order = Order {
+            exchange: instrument.instrument.exchange,
+            instrument: self.instrument,
+            strategy: self.id.clone(),
+            cid: ClientOrderId::default(),
+            side: Side::Buy,
+            state: RequestOpen {
+                kind: OrderKind::Market,
+                time_in_force: TimeInForce::ImmediateOrCancel,
+                price,
+                quantity: state.strategy.desired_quantity_to_buy,
+            },
+        };
+
+        (vec![], vec![order])
+    }
+}
+
+impl ClosePositionsStrategy for BuyAndHoldStrategy {
+    type State = EngineState;
+
+    fn close_positions_requests<'a>(
+        &'a self,
+        _state: &'a Self::State,
+        _filter: &'a InstrumentFilter,
+    ) -> (
+        impl IntoIterator<Item = Order<ExchangeIndex, InstrumentIndex, RequestCancel>> + 'a,
+        impl IntoIterator<Item = Order<ExchangeIndex, InstrumentIndex, RequestOpen>> + 'a,
+    )
+    where
+        ExchangeIndex: 'a,
+        AssetIndex: 'a,
+        InstrumentIndex: 'a,
+    {
+        // Here we can specify how should the positions be closed for our
+        // specific strategy.
+        (std::iter::empty(), std::iter::empty())
```

Comment:
In the absence of composition of strategies (ideal), there is a single strategy that implements all strategy traits (not ideal).

I've considered making this more composed, but it was non-trivial (at least without generating generic spaghetti). 
I have some ideas on that front though.

In the meantime, perhaps it's worth extracting the DefaultClosePositionsStrategy logic into a free standing pub function so it can easily be called by users defining their own Strategy.